Modules/Core/Common/include/itkDeref.h

```diff
@@ -0,0 +1,62 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         https://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkDeref_h
+#define itkDeref_h
+
+#include "itkMacro.h"
+#include <string>
+#include <typeinfo>
+
+namespace itk
+{
+
+/** \class DerefError
+ * Exception thrown when trying to dereference a null pointer.
+ * \ingroup ITKCommon
+ */
+class ITKCommon_EXPORT DerefError : public ExceptionObject
+{
+public:
+  // Inherit the constructors from its base class.
+  using ExceptionObject::ExceptionObject;
+
+  /** Runtime information support. */
+  itkTypeMacro(DerefError, ExceptionObject);
+};
+
+
+/** Dereferences the specified pointer (either "raw" or "smart"), when the pointer is not null. Throws a `DerefError`
+ * exception when the pointer is null. Aims to avoid undefined behavior from accidentally dereferencing a null pointer.
+ */
+template <typename TPointer>
+decltype(auto)
+Deref(const TPointer & ptr)
+{
+  if (ptr == nullptr)
+  {
+    throw DerefError(__FILE__,
```

Comment:
Is this file reported in the exception message, or the line where it is invoked?

Reply:
Yes indeed, both of them (source file and line number) are reported in the exception message.

Reply:
I meant, is file == ".../Modules/Core/Common/include/itkDeref.h", or file == ".../my/file/in/my/code/which/tried/to/deref/null.cpp"?

Reply:
The `itk::DerefError` exception message proposed by this PR will have something like "Modules/Core/Common/include/itkDeref.h" in there. Of course, it would have been nicer to have something like "my/file/in/my/code/which/tried/to/deref/null.cpp". I think that can be implemented most elegantly by C++20 [std::source_location](https://en.cppreference.com/w/cpp/utility/source_location) 😃. Until then, I think it's still nice for users to know the location where the `DerefError` is thrown, in "itkDeref.h", so that they can put a breakpoint at that location, when they are debugging the problem.

Reply:
Did you consider using the "itkSpecializedMessageExceptionMacro?

Reply:
Thanks @blowekamp, please check my [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/0f35aa8bd59c3793e1e694c7f3b0ec8c679b0617..f881d08b438ae1fe102bfa5eeae849389362c31f)!